File: client/android/src/main/java/io/codemagic/patch/CodemagicPatchPackageInstaller.kt
package io.codemagic.patch

import android.system.Os
import org.json.JSONObject
import java.io.File

internal class CodemagicPatchPackageInstaller(private val storage: CodemagicPatchStorage) {
  data class InstallContext(
    val packageHash: String,
    val binaryVersion: String,
    val deploymentKey: String,
    val runningPackageHash: String?,
  )

  fun install(context: InstallContext) {
    require(storage.isSafePackageHash(context.packageHash)) { "unsafe packageHash" }
    val record = storage.readJson("downloads/${context.packageHash}/download.json")
      ?: error("download record missing")
    val metadata = record.optJSONObject("metadata") ?: JSONObject()
    val current = storage.readState().current?.packageHash
    require(record.optString("package_hash") == context.packageHash) {
      "download record package hash mismatch"
    }
    val artifactType = record.optString("artifact_type")
    require(artifactType == "patch" || artifactType == "full_bundle") {
      "download record artifact type is invalid"
    }
    val expectedPayload = if (artifactType == "patch") "payload.patch.zst" else "payload.tar.zst"
    require(record.optString("payload") == expectedPayload) { "download payload path is invalid" }

    val tmpRoot = File(storage.root, "tmp/${context.packageHash}")
    val tmpContents = File(tmpRoot, "contents")
    tmpRoot.deleteRecursively()
    tmpRoot.mkdirs()

    val payload = File(storage.root, "downloads/${context.packageHash}/$expectedPayload")
    if (!payload.isFile) error("download payload missing")
    if (artifactType == "patch") {
      val base = record.optString("base_package_hash")
      val baseEligible = base.isNotBlank() &&
          base == context.runningPackageHash &&
          storage.metadataMatchesBinary(base, context.binaryVersion)
      val baseContents = storage.packageContentsDir(base)
      if (!baseEligible || !baseContents.isDirectory) {
        error("patch base package missing")
      }
      if (!NativeBundleOps.applyDirectoryPatch(
          baseContents.absolutePath,
          payload.absolutePath,
          tmpContents.absolutePath
      )) {
        error("patch application failed")
      }
    } else {
      if (!NativeBundleOps.extractBundle(payload.absolutePath, tmpContents.absolutePath)) {
        error("bundle extraction failed")
      }
    }

    if (!NativeBundleOps.validateContentsTree(tmpContents.absolutePath)) {
      tmpRoot.deleteRecursively()
      error("unsafe package contents")
    }

    val computedHash = NativeHashing.computePackageHash(tmpContents.absolutePath)
    if (computedHash != context.packageHash) {
      tmpRoot.deleteRecursively()
      error("package hash mismatch")
    }

    promotePackageContents(context.packageHash, tmpContents)
    tmpRoot.deleteRecursively()

    val updateJson = codemagicPatchUpdateJson(
      packageHash = context.packageHash,
      binaryVersion = context.binaryVersion,
      deploymentKey = context.deploymentKey,
      artifactType = artifactType,
      metadata = metadata,
      installedAt = CodemagicPatchUtil.currentIsoTimestamp(),
    )
    storage.writeJson("packages/${context.packageHash}/update.json", updateJson)
    storage.mutateState { state ->
      if (current != null && current != context.packageHash) {
        state.previous = CodemagicPatchPackagePointer(current)
      }
      state.pending = CodemagicPatchPackagePointer(context.packageHash)
      state.pendingStarted = null
    }
  }

  fun confirmPending() {
    storage.mutateState { state ->
      val pending = state.pending?.packageHash
      if (pending != null) {
        state.current = CodemagicPatchPackagePointer(pending)
        state.pending = null
        state.pendingStarted = null
      }
    }
  }

  fun stageEmbeddedRevert() {
    storage.mutateState { state ->
      state.current = null
      state.previous = null
      state.pending = null
      state.pendingStarted = null
    }
  }

  fun clearForTests() {
    storage.delete("packages")
    storage.delete("downloads")
    storage.delete("tmp")
    storage.delete("state")
  }

  private fun promotePackageContents(packageHash: String, tmpContents: File) {
    val packageRoot = File(storage.root, "packages/$packageHash")
    val contentsDir = File(packageRoot, "contents")
    packageRoot.mkdirs()

    if (contentsDir.exists()) {
      val existingValid =
          contentsDir.isDirectory &&
              NativeBundleOps.validateContentsTree(contentsDir.absolutePath) &&
              NativeHashing.computePackageHash(contentsDir.absolutePath) == packageHash
      require(existingValid) { "existing package contents are invalid" }
      return
    }

    storage.fsyncTree(tmpContents)
    Os.rename(tmpContents.absolutePath, contentsDir.absolutePath)
    storage.fsyncDirectory(packageRoot)
  }
}

internal fun codemagicPatchUpdateJson(
    packageHash: String,
    binaryVersion: String,
    deploymentKey: String,
    artifactType: String,
    metadata: JSONObject,
    installedAt: String,
): JSONObject =
    JSONObject()
      .put("package_hash", packageHash)
      .put("binary_version", binaryVersion)
      .put("deployment_key", deploymentKey)
      .put("label", metadata.optString("label", packageHash))
      .put("is_mandatory", metadata.optBoolean("isMandatory", false))
      .put("release_notes", nullableMetadataString(metadata, "releaseNotes"))
      .put("installed_at", installedAt)
      .put("source", artifactType)
      .put("signature_verified", metadata.optBoolean("signatureVerified", false))

private fun nullableMetadataString(metadata: JSONObject, key: String): Any =
    if (metadata.isNull(key)) JSONObject.NULL else metadata.optString(key)

File: client/ios/CodemagicPatchPackageInstaller.swift
import Foundation

final class CodemagicPatchPackageInstaller {
  struct InstallContext {
    let packageHash: String
    let binaryVersion: String
    let deploymentKey: String
    let runningPackageHash: String?
  }

  private let storage: CodemagicPatchStorage

  init(storage: CodemagicPatchStorage) {
    self.storage = storage
  }

  func install(_ context: InstallContext) throws {
    guard CodemagicPatchStorage.isSafePackageHash(context.packageHash),
          let record = storage.readJson("downloads/\(context.packageHash)/download.json") else {
      throw makeError("download record missing")
    }
    let metadata = record["metadata"] as? [String: Any] ?? [:]
    guard record["package_hash"] as? String == context.packageHash else {
      throw makeError("download record package hash mismatch")
    }
    guard let artifactType = record["artifact_type"] as? String,
          artifactType == "patch" || artifactType == "full_bundle" else {
      throw makeError("download record artifact type is invalid")
    }
    let expectedPayload = artifactType == "patch" ? "payload.patch.zst" : "payload.tar.zst"
    guard record["payload"] as? String == expectedPayload else {
      throw makeError("download payload path is invalid")
    }

    let tmpRoot = storage.url("tmp/\(context.packageHash)")
    let tmpContents = tmpRoot.appendingPathComponent("contents", isDirectory: true)
    try? FileManager.default.removeItem(at: tmpRoot)
    try FileManager.default.createDirectory(at: tmpRoot, withIntermediateDirectories: true)

    let payload = storage.url("downloads/\(context.packageHash)/\(expectedPayload)")
    guard FileManager.default.fileExists(atPath: payload.path) else {
      throw makeError("download payload missing")
    }
    if artifactType == "patch" {
      guard let base = record["base_package_hash"] as? String,
            base == context.runningPackageHash,
            storage.metadataMatchesBinary(packageHash: base, binaryVersion: context.binaryVersion),
            FileManager.default.fileExists(atPath: storage.packageContentsDir(base).path) else {
        throw makeError("patch base package missing")
      }
      try runArtifactOperation("patch application failed") {
        CodemagicPatchBundleOps.applyDirectoryPatch(
          withBaseDir: storage.packageContentsDir(base).path,
          patchPath: payload.path,
          outDir: tmpContents.path
        )
      }
    } else {
      try runArtifactOperation("bundle extraction failed") {
        CodemagicPatchBundleOps.extractBundle(payload.path, toOutDir: tmpContents.path)
      }
    }

    try runArtifactOperation("unsafe package contents") {
      CodemagicPatchBundleOps.validateContentsTree(tmpContents.path)
    }

    guard CodemagicPatchHashing.packageHash(atPath: tmpContents.path) == context.packageHash else {
      try? FileManager.default.removeItem(at: tmpRoot)
      throw makeError("package hash mismatch")
    }

    let packageRoot = storage.url("packages/\(context.packageHash)")
    try promotePackageContents(hash: context.packageHash, tmpContents: tmpContents, packageRoot: packageRoot)
    try? FileManager.default.removeItem(at: tmpRoot)

    try storage.writeJson("packages/\(context.packageHash)/update.json", [
      "package_hash": context.packageHash,
      "binary_version": context.binaryVersion,
      "deployment_key": context.deploymentKey,
      "label": metadata["label"] as? String ?? context.packageHash,
      "is_mandatory": metadata["isMandatory"] as? Bool ?? false,
      "release_notes": metadata["releaseNotes"] ?? NSNull(),
      "installed_at": CodemagicPatchUtil.currentIsoTimestamp(),
      "source": artifactType,
      "signature_verified": metadata["signatureVerified"] as? Bool ?? false
    ])
    let current = storage.readState().current?.packageHash
    try storage.mutateState { state in
      if let current = current, current != context.packageHash {
        state.previous = CodemagicPatchPackagePointer(packageHash: current)
      }
      state.pending = CodemagicPatchPackagePointer(packageHash: context.packageHash)
      state.pendingStarted = nil
    }
  }

  func confirmPending() throws {
    try storage.mutateState { state in
      if let pending = state.pending?.packageHash {
        state.current = CodemagicPatchPackagePointer(packageHash: pending)
        state.pending = nil
        state.pendingStarted = nil
      }
    }
  }

  func stageEmbeddedRevert() {
    try? storage.mutateState { state in
      state.current = nil
      state.previous = nil
      state.pending = nil
      state.pendingStarted = nil
    }
  }

  func clearForTests() {
    ["packages", "downloads", "tmp", "state"].forEach {
      storage.removeRelative($0)
    }
  }

  private func promotePackageContents(
    hash: String,
    tmpContents: URL,
    packageRoot: URL
  ) throws {
    let contents = packageRoot.appendingPathComponent("contents", isDirectory: true)
    try FileManager.default.createDirectory(at: packageRoot, withIntermediateDirectories: true)

    var isDirectory: ObjCBool = false
    if FileManager.default.fileExists(atPath: contents.path, isDirectory: &isDirectory) {
      let existingValid = isDirectory.boolValue &&
        CodemagicPatchBundleOps.validateContentsTree(contents.path) &&
        CodemagicPatchHashing.packageHash(atPath: contents.path) == hash
      guard existingValid else {
        throw makeError("existing package contents are invalid")
      }
      return
    }

    storage.fsyncTree(tmpContents)
    try FileManager.default.moveItem(at: tmpContents, to: contents)
    storage.fsyncDirectory(packageRoot)
  }

  private func runArtifactOperation(_ message: String, _ operation: () -> Bool) throws {
    if !operation() {
      throw makeError(message)
    }
  }

  private func makeError(_ message: String) -> NSError {
    NSError(domain: "CodemagicPatch", code: 2, userInfo: [NSLocalizedDescriptionKey: message])
  }
}
